src/abe/dippe.c

```diff
@@ -0,0 +1,613 @@
+    ECP_BN254 g1_x_UA_s[dippe->assump_size+1];
+
+    // Ci (pol->size x k+1)
+    cipher->Ci = (ECP_BN254*)malloc(sizeof(ECP_BN254) * pol->size * (dippe->assump_size+1));
+    for (unsigned int m=0; m<(pol->size); m++) {
+
+        // g1^(W_A_s)
+        for (unsigned int i=0; i<(dippe->assump_size+1); i++) {
+            ECP_BN254_inf(&(g1_W_A_s[i]));
+            for (unsigned int k=0; k<dippe->assump_size; k++) {
+                ECP_BN254_copy(&tmp_g1, &(pks[m]->g1_W_A[(dippe->assump_size*i)+k]));
+                BIG_256_56_from_mpz(tmp_big, s.vec[k]);
+                ECP_BN254_mul(&tmp_g1, tmp_big);
+                ECP_BN254_add(&(g1_W_A_s[i]), &tmp_g1);
+            }
+        }
+
+        // g1^(x_UA_s)
+        for (unsigned int i=0; i<(dippe->assump_size+1); i++) {
+            ECP_BN254_inf(&(g1_x_UA_s[i]));
+            for (unsigned int k=0; k<dippe->assump_size; k++) {
+                ECP_BN254_copy(&tmp_g1, &(dippe->g1_UA[(dippe->assump_size*i)+k]));
+                BIG_256_56_from_mpz(tmp_big, s.vec[k]);
+                ECP_BN254_mul(&tmp_g1, tmp_big);
+                ECP_BN254_add(&(g1_x_UA_s[i]), &tmp_g1);
+            }
+            BIG_256_56_from_mpz(tmp_big, pol->vec[m]);
+            ECP_BN254_mul(&(g1_x_UA_s[i]), tmp_big);
+        }
+
+        // add both matrices
+        for (unsigned int i=0; i<(dippe->assump_size+1); i++) {
+            ECP_BN254_copy(&(cipher->Ci[(m*(dippe->assump_size+1))+i]), &(g1_x_UA_s[i]));
+            ECP_BN254_add(&(cipher->Ci[(m*(dippe->assump_size+1))+i]), &(g1_W_A_s[i]));
+        }
+    }
+
+    // C_prime
+    FP12_BN254_one(&(cipher->C_prime));
+    for (unsigned int m=0; m<pol->size; m++) {
+        for (unsigned int k=0; k<dippe->assump_size; k++) {
+            BIG_256_56_from_mpz(tmp_big, s.vec[k]);
+            FP12_BN254_pow(&tmp_gt, &(pks[m]->gt_alpha_A[k]), tmp_big);
+            FP12_BN254_mul(&(cipher->C_prime), &tmp_gt);
+        }
+    }
+
+    // blind message
+    FP12_BN254_mul(&(cipher->C_prime), msg);
+
+    // cleanup
+    cfe_vec_free(&s);
+
+    return CFE_ERR_NONE;
+}
+
+/**
+ * Used with an authority's secret key to create a new user secret key for a given attribute vector
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param usk A pointer to a cfe_dippe_user_secret_key struct; Represents the user secret key that is about to be populated
+ * @param usk_id Index of the given attribute vector for which a user secret key will be created
+ * @param pks An array containing references to cfe_dippe_pub_key structs
+ * @param pks_len Length of the public key array
+ * @param attrs A reference to a mpz_t struct; Represents the attribute vector
+ * @param gid String that represents a unique user; Required for collusion prevention
+ * @return Error code
+ */
+cfe_error cfe_dippe_keygen(cfe_dippe *dippe, cfe_dippe_user_secret_key *usk, unsigned int usk_id, cfe_dippe_pub_key *pks[], unsigned int pks_len, cfe_dippe_sec_key *sk, mpz_t *attrs, char gid[]) {
```

Comment:
Choosing `attrs` to be a `mpz_t` seems like a strange choice. As far a I see (correct me if I am wrong) you are just changing a string to `mpz_t` and then back to string. I would suggest to have something like an array of owned attributes, for example `int attrib[] = {0,3,4}` indicating atribute 0, 3 and 4 are owned. We did it this way in FAME scheme.